TOT-55 : [FEAT] 백엔드 카카오 로그인 기능 개발

snowcrab382 opened the pull request:
## PR전 체크리스트
> 1. PR 및 Commit title을 형식에 맞게 작성했나요(e.g. TOT-123 : [TYPE])
네
> 2. Reviewers가 명확하게 지정됐나요?
네
> 3. Assignees가 명확하게 지정됐나요?
네


## 📝 작업 내용
- 카카오 로그인 로직 구현
- jwt 발급,검증,예외처리 필터 구현

### 스크린샷
- 카카오 로그인 시퀀스 다이어그램
![image](https://github.com/trip-or-treat/trip-or-treat/assets/108510951/a4827fdc-c9f3-4462-bba3-57981d00dc1e)

- jwt 검증 흐름도
![image](https://github.com/trip-or-treat/trip-or-treat/assets/108510951/72d6df8c-2d3b-4875-b16a-2d468eec2005)

## 안내 사항
- 현재 백엔드 기능이 업데이트되어야 프론트쪽 로그인 개발이 원활히 진행될 것 같다고 하셔서 조금 이르게 PR 올리게 되었습니다.
- 리팩토링, 테스트코드 작성 등은 프론트 개발이 완료된 후 기능 점검 및 버그 수정과 함께 작업하여 추가하겠습니다.
- 일반적인 API 흐름과는 조금 다른 부분이 있기 때문에 로그와 주석을 일부 작성하였습니다. 올려드린 이미지와 함께 흐름을 따라가보시면 이해하시는데에 조금 더 도움이 될 것 같습니다.

## 💬 리뷰 요구사항

## 참고자료

Commit:
TOT-55 : [CONFIG] jwt 라이브러리 의존성 주입

Commit:
TOT-55 : [FEAT] 로그인 컨트롤러 구현
- 클라이언트로부터 카카오 인가코드를 발급받아 로그인 로직 수행

Commit:
TOT-55 : [FEAT] 인증 서비스 핵심 로직 구현
- 요청으로 전달받은 카카오 인가코드를 이용해 카카오 엑세스토큰 발급
- 발급받은 토큰으로 사용자 정보 획득
- 사용자 email 기준으로 DB에 존재하면 반환, 존재하지 않으면 저장 후 반환
- 사용자ID를 토큰정보에 담아 토큰 생성
- 응답에 토큰 저장

Commit:
TOT-55 : [FEAT] 서버측 토큰 발급 클래스 구현
- jwt 발급, 재발급 로직
- 응답 헤더에 액세스토큰 저장 로직
- 응답 쿠키에 리프레시토큰 저장 로직
- 클라이언트 요청에서 액세스토큰과 리프레시토큰 추출 로직
- jwt 유효성 검증 로직
- 토큰에서 인증에 필요한 유저정보 파싱 로직

Commit:
TOT-55 : [FEAT] jwt 검증 및 예외처리 필터 구현
- whitelist와 일치하는 uri인 경우 jwt 검증로직 스킵(인증이 필요없는 API)
- 액세스, 리프레시 토큰 검증 후 통과하면 컨트롤러로 요청 전달
- 검증 중 예외발생시 예외처리 필터로 예외 전달
- 필터에서 발생한 예외는 exceptionHandler로 처리가 불가하므로 자체 handler 구현

Commit:
TOT-55 :  [FEAT] 로그인 성공메세지 추가

Commit:
TOT-55 : [FEAT] 토큰에서 추출한 사용자정보 컨트롤러 전달로직 구현
- argumentResolver를 활용해 토큰에 포함된 사용자정보 추출
- Auth 커스텀 애노테이션 생성하여 해당 애노테이션을 통해 사용자정보 전달

Commit:
TOT-55 :  [CONFIG] jwt, 카카오 관련 설정 추가

Commit:
TOT-55 :  [REFACTOR] 유틸성 클래스파일 패키지 분리

Commit:
TOT-55 :  [FIX] 로그인 관련 암호화정보 불일치 해결

Commit:
TOT-55 :  [REFACTOR] 로그인 API 요청 방식 수정

MisaSohee commented:
올려주신 다이어그램과 흐름도, commit 메세지 덕분에 코드를 이해하는데 많이 도움 되었습니다!
테스트코드는 추후에 올리신다 하셨으니 승인이 되면 이대로 merge 하시는 걸까요??

Commit:
TOT-55 :  [FIX] 프론트 작업을 위한 임시 설정 변경

Commit:
TOT-55 : [FIX] @WebMvcTest를 사용하는 컨트롤러 테스트 실패 문제 해결
- @WebMvcTest가 테스트를 위해 의존 주입하는 Filter, WebMvcConfigurer, Resolver 등의
  클래스 파일들을 생략하도록 설정

snowcrab382 commented:
> 올려주신 다이어그램과 흐름도, commit 메세지 덕분에 코드를 이해하는데 많이 도움 되었습니다! 테스트코드는 추후에 올리신다 하셨으니 승인이 되면 이대로 merge 하시는 걸까요??

네 맞습니다! 이후 리팩토링 과정을 진행하면서 로그와 같이 불필요한 부분은 삭제할 예정입니다.

snowcrab382 committed:
Merge branch 'develop' into FEAT/TOT-55

honee8583 commented on backend/src/main/java/com/triportreat/backend/auth/domain/KakaoLoginRequestDto.java:
```diff
@@ -0,0 +1,9 @@
+
+import lombok.Getter;
+
+@Getter
+public class KakaoLoginRequestDto {
+
+    private String code;
+}
```
이 Dto클래스는 어디서 사용되는건가요??

snowcrab382 replied:
아 기존에 카카오 인가코드를 받는 용도로 Dto를 생성했었는데 안쓰게 되어서 삭제해야되는걸 깜빡했네요 ㅠ 삭제하도록 하겠습니다!

honee8583 reviewed (approved):
고생하셨습니다! 스프링 시큐리티에 있는 필터를 상속받아 구현하지 않고 일반 필터로 구현한 이유가 궁금합니다!

snowcrab382 commented:
> 스프링 시큐리티에 있는 필터를 상속받아 구현하지 않고 일반 필터로 구현한 이유가 궁금합니다!

처음에는 시큐리티를 학습해서 적용하려고 했었는데, 막상 로그인 흐름을 파악하고보니 서블릿 필터로도 충분히 구현이 가능할 것 같아서 강의에서 들었던 필터도 사용해볼 겸 사용하게되었습니다. 이 경험 이후에 시큐리티를 사용한 코드로 리팩토링해보면 조금 더 깊이 있게 학습이 가능할 것 같다는 생각입니다!

snowcrab382 committed:
[FIX] 불필요한 요청dto 삭제
